bundles/nl.asml.matala.product/src/nl/asml/matala/product/validation/ProductValidator.xtend

```diff
@@ -30,6 +49,115 @@ class ProductValidator extends AbstractProductValidator {
     }
     
     
+    /**
+     * Check the duplication of variables in Inputs
+     */
+     
+     @Check
+     def checkInputDuplication(Block b){ 
+            var HashSet<String> existingInputs = new HashSet<String>();
+            for (input : b.invars) {
+                if (!existingInputs.add(input.name)) {
+                    error("There exists a variable name in inputs with the same name" ,  ProductPackage.Literals.BLOCK__INVARS)  
+                }
+            }
+        }
+
+    /**
+     * Check the duplication of variables in Outputs
+     */
+     @Check
+     def checkOutputDuplication(Block b){
+           var HashSet<String> existingOutputs = new HashSet<String>();
+            for (output : b.outvars) {
+                if (!existingOutputs.add(output.name)) {
+                    error("There exists a variable name in outputs with the same name" ,  ProductPackage.Literals.BLOCK__OUTVARS)  
+                }
+            }
+        }
+    
+      /**
+     * Check the duplication of variables in Local
+     */
+     @Check
+     def checkLocalDuplication(Block b){
+        var HashSet<String> existingLocalvars = new HashSet<String>();
+            for (localvars : b.localvars) {
+                if (!existingLocalvars.add(localvars.name)) {
+                    error("There exists a variable name in local with the same name" ,  ProductPackage.Literals.BLOCK__LOCALVARS)  
+                }
+            }
+        }
+
+    
+    
+        
+    /**
+     * All variables occurring in a guard expression must have been defined as inputs of an action
+     */
+    @Check
+    def preventIlligalVariableAccess(Function f){
+          var allVariables = new HashSet<String>
+          var inputes = new HashSet<String>
+          for(update : f.updates){
+            if(update.guard !== null){
+               // Get all variables in a guard
+                val allExprVariables = EcoreUtil2::getAllContentsOfType(update.guard, ExpressionVariable)
```

Comment:
Xtend has a very powerfull query language built-in, so you could re-write building the allExprVariables as:

`var allVariables = update.guard.eAllContents.filter(ExpressionVariable).map[variable.name].toSet`